src/CodeGen_PTX_Dev.cpp

```diff
@@ -791,6 +809,256 @@ bool CodeGen_PTX_Dev::supports_atomic_add(const Type &t) const {
+            const Cast *cast = matches[1].as<Cast>();
+
+            if (matmul_load && cast) {
+                // Check if the load is loading from the same place where the store is storing
+                // i.e. if this is an update operation
+                if (matmul_load->name == store->name && equal(matmul_load->index, store->index)) {
+                    // Yes, this is an update operation, now let's check cast to see if it is a matmul that can be converted
+                    // to wmma intrinsic
+
+                    // float32((A[((A.stride.1*t34) - t33) + matmul$1.s1.k$x]*B[(B.stride.1*matmul$1.s1.k$x) + t35]))
+                    const Expr wild_f16x = Variable::make(Float(16), "*");
+                    const Expr mul_pattern = wild_f16x * wild_f16x;
+                    if (expr_match(mul_pattern, cast->value, matches)) {
+                        // A[((A.stride.1*t34) - t33) + matmul$1.s1.k$x]
+                        const Load *load_a = matches[0].as<Load>();
+                        // B[(B.stride.1 * matmul$1.s1.k$x) + t35]
+                        const Load *load_b = matches[1].as<Load>();
+
+                        if (load_a && load_b) {
+                            const Expr wild_i32x = Variable::make(Int(32), "*");
+                            const Expr k_var_expr = Variable::make(Int(32), k_var_name);
+                            // Check if the reduction domain is being used in both A and B to
+                            // validate the matrix multiply
+                            const Expr load_a_pattern = wild_i32x + wild_i32x;
+                            const Expr load_b_pattern = wild_i32x * wild_i32x + wild_i32x;
+                            vector<Expr> matches_a, matches_b;
+                            const bool match_a = expr_match(load_a_pattern, load_a->index, matches_a);
+                            const bool match_b = expr_match(load_b_pattern, load_b->index, matches_b);
+                            if (match_a && match_b) {
+                                // Check if the k_var_name is present in the expressions for load_a and load_b
+                                const Variable *load_a_var_1 = matches_a[0].as<Variable>();
+                                const Variable *load_a_var_2 = matches_a[1].as<Variable>();
+                                const Variable *load_b_var_1 = matches_b[0].as<Variable>();
+                                const Variable *load_b_var_2 = matches_b[1].as<Variable>();
+
+                                // TODO: Can this checks be improved?
+                                const bool load_a_ok = (load_a_var_1 && load_a_var_1->name == k_var_name) || (load_a_var_2 && load_a_var_2->name == k_var_name);
+                                const bool load_b_ok = (load_b_var_1 && load_b_var_1->name == k_var_name) || (load_b_var_2 && load_b_var_2->name == k_var_name);
+
+                                if (load_a_ok && load_b_ok) {
```

Comment:
This could be triggered by the pattern A(x, k) * B(x, k), right? There doesn't seem to be a check that the reduction var is used in different dimensions for A and B

Reply:
I accidently marked it as resolved. Sorry, I will check this expression now

Reply:
@rootjalex This particular expression you mentioned won't be matched the first value of `matches_a` won't be a variable.

Reply:
Hmm but the patterns don't use `k_var_expr`, so I'm not confident about that - but I guess the bug might be in the pattern definitions.

Reply:
When I first wrote this, I had a pattern like this

```c++
const Expr load_a_pattern = wild_i32x + Variable::make(k_var_name, Int(32));
const Expr load_b_pattern = wild_i32x * Variable::make(k_var_name, Int(32)) + wild_i32x;
```

But then I found out that `expr_match` expects the `k_var` to be on the right hand side of the `Add` and `Mul` respectively. This is not always the case, so I replaced it with a wild card expression and I check for the `k_var_name` on lines 874 and 875

Reply:
Ah gotcha, thanks for the explanation. I incorrectly assumed that the matcher would catch that case, my apologies.

Reply:
If you have any suggestion to improve this match please let me know